### PR TITLE
Add S3 Storage Backend

### PR DESCRIPTION
**Pull request by nupplaphil**
It's working :-)

I needed to fix the akeeba library as well (will create a PR for them), but at least I didn't have to include the official AWS library, which contains more than **8.300** files! I thought that would be insane for our specific use case.

I tested the S3 Backend with a Amazon S3 bucket and it worked well
- Create tested
- Update (rotating) tested
- Delete tested

I'd say it's ready :-)

btw. the tests will get fixed after #1238

### CODE REVIEW

File: s3_storage/composer.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	{
+	  "name": "Philipp Holzer",
+	  "email": "admin@philipp.info",
+	  "homepage": "https://blog.philipp.info",
+	  "role": "Developer"
+	}
+  ],
+  "require": {
+	"php": ">=7.0",
+    "akeeba/s3": "^2.0"
```
Comment (MrPetovan) [inline s3_storage/composer.json:15]: Code standards: Please normalize white space in the file.

File: s3_storage/tests/S3Test.php.bak [hunk 2]
```diff
@@ -0,0 +1,121 @@
+<?php
```
Comment (MrPetovan) [inline s3_storage/tests/S3Test.php.bak]: Is this file `.bak` meant to be committed?

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: s3_storage/vendor/akeeba/s3/README.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+This code was originally based on [S3.php written by Donovan Schonknecht](http://undesigned.org.za/2007/10/22/amazon-s3-php-class) which is available under a BSD-like license. This repository no longer reflects the original author's work and should not be confused with it.
+
+This software is distributed under the GNU General Public License version 3 or, at your option, any later version published by the Free Software Foundation (FSF). In short, it's "GPLv3+".
+
+## Important note about version 2
+
+Akeeba Amazon S3 Connector version 2 has dropped support for PPH 5.3 to 7.0 inclusive. It is only compatible with PHP 7.1 or later, up to and including PHP 8.0.
+
+The most significant change in this version is that all methods use scalar type hints for parameters and return values. This _may_ break existing consumers which relied on implicit type conversion e.g. passing strings containing integer values instead of _actual_ integer values.
```
Comment (MrPetovan) [inline s3_storage/vendor/akeeba/s3/README.md:17]: Good news for these consumers, scalar type hints also do implicit conversion from strings to integers!

File: s3_storage/vendor/akeeba/s3/src/Acl.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @package   akeebaengine
+ * @copyright Copyright (c)2006-2020 Nicholas K. Dionysopoulos / Akeeba Ltd
+ * @license   GNU General Public License version 3, or later
+ */
+
+namespace Akeeba\Engine\Postproc\Connector\S3v4;
+
+// Protection against direct access
+defined('AKEEBAENGINE') or die();
```
Comment (MrPetovan) [inline s3_storage/vendor/akeeba/s3/src/Acl.php:13]: Huh? What's the point of this in all the files?

**Review by MrPetovan — APPROVED** [review]

**Comment by AlfredSK** [comment]
I'll check this out with my test node mordor.social and a Minio server. Thank you!

**Comment by nupplaphil** [comment]
> I'll check this out with my test node mordor.social and a Minio server. Thank you!

I think you have to use v2/legacy flag, afaik

File: s3_storage/vendor/akeeba/s3/src/Configuration.php [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+			if (strpos($this->endpoint, 'amazonaws.com') !== false)
+			{
+				$this->setUseLegacyPathStyle(false);
+			}
+
+		} else {
+			if (empty($this->getRegion())) {
+				$this->setRegion('us-east-1');
+			}
+		}
```
Comment (MrPetovan) [inline s3_storage/vendor/akeeba/s3/src/Configuration.php:221]: Why not 
```php
		} elseif (empty($this->getRegion())) {
			$this->setRegion('us-east-1');
		}
```

**Pull request merged**